docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,325 @@
+  1. This is very dangerous to do automatically and is probably not realistic at this time:
+    1. Requires automated rollback viability analysis/trust.
+    2. Requires logic to prevent a chronically failing hook that spans historical deployments from causing unending rollbacks.
+
+This proposal prescribes the use of option 1 as being the simplest starting point for the hook API.
+
+Failed mandatory post-deployment hooks are more challenging:
+
+1. The deployment has most likely already been rolled out and made live by the strategy.
+2. Deleting the deployment is no longer safe due to 1.
+3. Rollback is necessary and subject to the same challenges presented above.
+
+Due to the complexities of automated rollback, this proposal limits the scope of failure handling for post-deployment hooks: post-deployment hooks cannot be considered mandatory at this time. This limitation may be lifted in the future by an separate proposal.
+
+##### Hook failure reporting
+
+When a deployment hook fails:
+
+1. An error is logged via the global error handler.
+2. The hook status is available as an annotation on the deployment.
+
+More reporting capabilities could be addressed in a future proposal.
+
+
+### Deployment hooks API
+
+The `DeploymentStrategy` gains a new `Lifecycle` field:
+
+```go
+type DeploymentStrategy struct {
+  // Type is the name of a deployment strategy.
+  Type DeploymentStrategyType `json:"type,omitempty"`
+  // CustomParams are the input to the Custom deployment strategy.
+  CustomParams *CustomDeploymentStrategyParams `json:"customParams,omitempty"`
+  // Lifecycle provides optional hooks into the deployment process.
+  Lifecycle *Lifecycle `json:"lifecycle,omitempty"`
+}
+```
+
+```go
+// Lifecycle describes actions the system should take in response to
+// deployment lifecycle events. The deployment process blocks while
+// executing lifecycle handlers. A HandleFailurePolicy determines what
+// action is taken in response to a failed handler.
+type Lifecycle struct {
+  // Pre is called immediately before the deployment strategy executes.
+  Pre *Handler `json:"pre,omitempty"`
+  // Post is called immediately after the deployment strategy executes.
+  // NOTE: AbortHandlerFailurePolicy is not supported for Post.
+  Post *Handler `json:"post,omitempty"`
+}
+```
+
+Each lifecycle hook is implemented with a `Handler`:
+
+```go
+// Handler defines a specific deployment lifecycle action.
+type Handler struct {
+  // ExecNewPod specifies the action to take.
+  ExecNewPod *ExecNewPodAction `json:"execNewPod,omitempty"`
+  // FailurePolicy specifies what action to take if the handler fails.
+  FailurePolicy HandlerFailurePolicy `json:"failurePolicy"`
+}
+```
+
+The first handler implementation is pod-based:
+
+```go
+// ExecNewPodAction runs a command in a new pod based on the specified
+// container which is assumed to be part of the deployment template.
+//
+// TODO: describe things that are/aren't included (e.g. volumes)
```

Comment:
Yeah, since the spec includes a tag which currently refers to an image ID, and soon will be a spec which contains the new digest hash.